gdal/ogr/ogrsf_frmts/ngw/gdalngwdataset.cpp

```diff
@@ -0,0 +1,911 @@
+            {
+                // GetExtent from parent.
+                OGREnvelope stExtent;
+                std::string osParentId = oRoot.GetString("resource/parent/id");
+                bool bExtentResult = NGWAPI::GetExtent(osUrl, osParentId,
+                    papszHTTPOptions, 3857, stExtent);
+
+                if( !bExtentResult )
+                {
+                    // Set full extent for EPSG:3857.
+                    stExtent.MinX = -20037508.34;
+                    stExtent.MaxX = 20037508.34;
+                    stExtent.MinY = -20037508.34;
+                    stExtent.MaxY = 20037508.34;
+                }
+
+                CPLDebug("NGW", "Raster extent is: %f, %f, %f, %f",
+                    stExtent.MinX, stExtent.MinY,
+                    stExtent.MaxX, stExtent.MaxY);
+
+                int nEPSG = 3857;
+                // Get parent details. We can skip this as default SRS in NGW is 3857.
+                if( osResourceType == "wmsclient_layer" )
+                {
+                    nEPSG = oRoot.GetInteger("wmsclient_layer/srs/id", nEPSG);
+                }
+                else
+                {
+                    CPLJSONDocument oResourceReq;
+                    bResult = oResourceReq.LoadUrl( NGWAPI::GetResource( osUrl,
+                        osResourceId ), papszHTTPOptions );
+
+                    if( bResult )
+                    {
+                        CPLJSONObject oParentRoot = oResourceReq.GetRoot();
+                        if( osResourceType == "mapserver_style" ||
+                            osResourceType == "qgis_vector_style" )
+                        {
+                            nEPSG = oParentRoot.GetInteger("vector_layer/srs/id", nEPSG);
+                        }
+                        else if( osResourceType == "raster_style")
+                        {
+                            nEPSG = oParentRoot.GetInteger("raster_layer/srs/id", nEPSG);
+                        }
+                    }
+                }
+
+                // Create raster dataset.
+                std::string osRasterUrl = NGWAPI::GetTMS(osUrl, osResourceId);
+                const char *pszConnStr = CPLSPrintf("<GDAL_WMS><Service name=\"TMS\">"
+            "<ServerUrl>%s</ServerUrl></Service><DataWindow>"
+            "<UpperLeftX>-20037508.34</UpperLeftX><UpperLeftY>20037508.34</UpperLeftY>"
+            "<LowerRightX>20037508.34</LowerRightX><LowerRightY>-20037508.34</LowerRightY>"
+            "<TileLevel>%d</TileLevel><TileCountX>1</TileCountX>"
+            "<TileCountY>1</TileCountY><YOrigin>top</YOrigin></DataWindow>"
+            "<Projection>EPSG:%d</Projection><BlockSizeX>256</BlockSizeX>"
+            "<BlockSizeY>256</BlockSizeY><BandsCount>%d</BandsCount>"
+            "<Cache><Type>file</Type><Expires>%d</Expires><MaxSize>%d</MaxSize>"
+            "</Cache><ZeroBlockHttpCodes>204,404</ZeroBlockHttpCodes></GDAL_WMS>",
+                osRasterUrl.c_str(),
```

Comment:
you should CPLEscapeString( CPLES_XML ) osRasterUrl

Reply:
Fix per 4579aeab927d70ddab6142340bf5550c6789b42e